### ISSUE TITLE
GDScript: add broadened support for operator '==' in comparision with Object vs primite/built-in types

### ISSUE THREAD

**Issue by just-like-that** [godotengine/godot-proposals#5101]
### Describe the project you are working on

2D-based games - any will do.

### Describe the problem or limitation you are having in your project

I frequently declare my own classes.

For example I've a tri-state type class or spatial types which I use in all my projects.

The TriState class supports its own type as well as booleans for comparison purposes
therefore I've a method:
 
```func equals(other) -> bool```

As GDScript does NOT allow method overloading with different parameter signatures.

The argument other is as open as necessary - it's NOT statically typed. 

As such I've to decide in the method body itself what I'm dealing with, like so:

```
	func equals(other) -> bool:
		return typeof(other) == TYPE_OBJECT && self == other
	###
```
Above example is just the most basic variation of this. Usually there are different code paths for dealing with different types.

This needs boilerplate code:
typeof(other) == TYPE_OBJECT

otherwise the second condition results in error:

Invalid operands 'Object' and 'bool' in operator '=='.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I propose to implement support for comparison between type Object and primitive/built-in types.

The message says it all: "Invalid operands 'Object' and 'bool' in operator '=='."
As such an object hence custom type will never be equal to a primitive/built-in type.

It's an instance to an object of a certain type. So I know it's not a primitive or built-in type.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

That said just return false which leads to the following code:
```
	func equals(other) -> bool:
		return self == other
	###
```

The logical equality operator should include the type check against Object implicitly
as everything which is NOT a reference to self will never be equal hence true.

From a programmer point of view we get rid of the redundant check against TYPE_OBJECT:
 typeof(other) == TYPE_OBJECT

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It's a GDScript lang feature and has to be handled by the programmer explicitly currently.

### Is there a reason why this should be core and not an add-on in the asset library?

It's core coz it's a GDScript language feature.

**Comment by YuriSizov** [comment]
> Always return false as such an object hence custom type will never be equal to a primitive/built-in type.

This would hide typing errors from developers with no alternative way to detect them. You having to add an extra check means you understand what you're doing and that other data types can be fed to the conditional. Explicitness over implicitness and coercion.

**Comment by just-like-that** [comment]
> This would hide typing errors from developers with no alternative way to detect them.

If I want to check explicitly I still may like so:

```
func equals(other) -> bool:
	if !other is Object : return false
	return other == self
###
```

which is still compatible with primitive and other built-in types.

Or do I miss something?

**Comment by just-like-that** [comment]
A quick test on runtime behavior revealed that the check:
- other is Object && other == self

seems to be faster than:
- typeof(other) == TYPE_OBJECT && other == self

To me it looks like a tautology.
https://en.wikipedia.org/wiki/Tautology_(logic)

I compare a non-object to an object (self) that's always false.

**Comment by YuriSizov** [comment]
> Or do I miss something?

Yes, you miss the fact that you need to manually think of the checks beforehand to catch this, instead of it failing on a type mismatch by default and informing you of a typing issue.

**Comment by KoBeWi** [comment]
Isn't this a duplicate of #4571?

**Comment by zinnschlag** [comment]
Not exactly. The proposal you linked also includes a special case for int and float comparison (or rather the removal of a special case).

**Comment by KoBeWi** [comment]
Ah right. Literally everyone who supported that proposal said that it's good without the float part.

**Comment by just-like-that** [comment]
One more thing:

- This line leads to a compile-time error:
`var other : int = null`
Understood. The assigned value type (null) doesn't match the variable's type (int).

- At the same time this works in runtime:
```
prints("5 == null", 5 == null) # works in runtime - prints False
prints("5 != null", 5 != null) # works in runtime - prints True
```

- The below two lines lead to an error:
```
print(5 == self) # error at compile time: Invalid operand types ('int' and 'self') to operator '=='.
print(5 != self) # error at compile time: Invalid operand types ('int' and 'self') to operator '=='.
```

Keyword null is the representation of a void instance. It's a null reference. 
As such an instance and null are counterparts. Both are meant for object types.
Keyword self is just the access to a valid reference from inside the class.

- Now the following only works if other is of the same type as 5 (int) OR null (!) otherwise an error is thrown.
```
prints("5 == other", 5 == other) # works in runtime if other is null - prints False
prints("5 != other", 5 != other) # works in runtime if other is null - prints True
```

This leads to the following question:

- Why is there a different treatment of keywords null and self as both represent an instance of type Object?

From an outside perspective it looks like an inconsistency.

In a line like `other == self` similar to `other == null` it's already known at compile-time that other has to be of type Object.

Therefore I propose to have at least a compiler-warning so that others are also aware of that fact and adapt their source accordingly.

Even more so as a self-check is a common pattern in other programming languages like Java.

Side note:
It would also help to support method signature overloading with different types like so ;):
`equals(val: <MyOwnClassType>) and equals(val: bool)`
But that's on a completely different level.
